### DISCUSSION TITLE
Github project expansion

### DISCUSSION THREAD

**Discussion by dpavicic** [thomhurst/ModularPipelines#436] [Ideas]
Hey Tom,
I was fiddling with ModularPipelines.GitHub to add support for the (auto authenticated) ```IGitHubClient``` and GitHub repository information regarding owner, repo name, etc.

Here's the commit:
https://github.com/six-tech/ModularPipelines/commit/4bdf3393dbf49c17098f926c86016d7b190f62df

The GitHub API is now extended with Client and RepositryInfo, so it looks like this:

context.GitHub().Client <- gets the reference to IGitHubClient and full OctoKit API
context.GitHub().RepositoryInfo.Owner (and other props)

What do you think about this approach?

Some benefits:
- It exposes auto-authenticated IGitHubClient so all APIs of OctoKit are at the disposal without any effort (needs GITHUB_TOKEN in env as usual)
- It exposes the GitHubRepositoryInfo object with some useful properties that are frequently used with the IGitHubClient (owner, repo name)

Some downs:
- The project now depends on ModularPipelines.Git (needed the root path, but if that's a problem, I can find the .git directory manually without this)
- The project now depends on OctoKit 10.0.0 and Lib2GitSharp. OctoKit is needed for the client, Lib2GitSharp, to be able to extract remote repo info. I tried with git commands, but they return Task, so this complicates what I need to do in the constructor.

**Comment by thomhurst** [comment]
Hey again @dpavicic !

Having the Octo client seems useful. I've ended up using it in a few other projects myself!

I don't see an issue with GitHub depending on Git. If you're using GitHub, you're using Git surely!

I've added a couple of comments around running async code to initialise a class. Because I think I'd like to avoid a dependency on Lib2GitSharp. 
Also maybe opening up other ways to pass in a token. Might also be worth starting to document this stuff for future users too if we go down this route. There's a docs folder which has mark up files, which dynamically builds the documentation site. Would you mind explaining any setup a user might want to do if we take this approach?

Let me know what you think.

Thanks 😄

**Reply by dpavicic** [reply]
Hey, here's the update:
- Add GitHub options with access token
- Implement IIntializable for async init of GitHubRepositoryInfo
- Add Url property to IGitHubRepositoryInfo
- Use IGit the get remote repo data and remove LibGit2Sharp
- Remove unneeded tuple for owner, repo
- Implement Lazy<IGitHubClient>

https://github.com/six-tech/ModularPipelines/commit/27db73469b5f9bf43c2fe7b3b58697ff2569ba97

Seems I covered everything we talked about; and it seems it's ok.
Here's the small example:

```csharp
public class GitHubTestModule : Module
{
    protected override async Task<IDictionary<string, object>?> ExecuteAsync(
        IPipelineContext context, 
        CancellationToken cancellationToken)
    {
        var info = context.GitHub().RepositoryInfo;
        var user = await context.GitHub()
            .Client
            .User
            .Get(info.Owner);
        
        context.Logger.LogInformation("User location: {Tag}", user.Location);
        
        await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken);
        return null;
    }
}
```

Regarding the docs; sure, I will write something when all of this is ok, stable and ok to merge.

**Comment by thomhurst** [comment]
Cheers looks good! Just added two more places where I think we should make lazy

**Reply by dpavicic** [reply]
Went over the things we talked about:
https://github.com/six-tech/ModularPipelines/commit/e5d93450f85850c938cae2d893a82b2aff40833a

**Reply by thomhurst** [reply]
Looks good - Thanks! Feel free to raise a PR
